contexts/ProposalMarketsContext.tsx

```diff
@@ -0,0 +1,602 @@
+            ]);
+            const failOrders = await openbook.account.openOrdersAccount.all([
+                { memcmp: { offset: 8, bytes: owner.toBase58() } },
+                { memcmp: { offset: 40, bytes: proposal.account.openbookFailMarket.toBase58() } },
+            ]);
+            return passOrders
+                .concat(failOrders)
+                .sort((a, b) => (a.account.accountNum < b.account.accountNum ? 1 : -1));
+        };
+
+        const orders = await client.fetchQuery({
+            queryKey: [`fetchProposalOpenOrders-${proposal?.publicKey}`],
+            queryFn: () => fetchProposalOpenOrders(),
+            staleTime: 1_000,
+        });
+        setOrders(orders ?? []);
+    }, [openbook, proposal]);
+
+    useEffect(() => {
+        if (proposal && wallet.publicKey) {
+            fetchOpenOrders(wallet.publicKey);
+        }
+    }, [markets, fetchOpenOrders, proposal]);
+
+    useEffect(() => {
+        if (!markets && proposal) {
+            fetchMarketsInfo();
+        }
+    }, [markets, fetchMarketsInfo, proposal]);
+
+    const orderBookObject = useMemo(() => {
+        const getSide = (side: LeafNode[], isBidSide?: boolean) => {
+            if (side.length === 0) {
+                return null;
+            }
+            const parsed = side
+                .map((e) => ({
+                    price: e.key.shrn(64).toNumber(),
+                    size: e.quantity.toNumber(),
+                }))
+                .sort((a, b) => a.price - b.price);
+
+            const sorted = isBidSide
+                ? parsed.sort((a, b) => b.price - a.price)
+                : parsed.sort((a, b) => a.price - b.price);
+
+            const deduped = new Map();
+            sorted.forEach((order) => {
+                if (deduped.get(order.price) === undefined) {
+                    deduped.set(order.price, order.size);
+                } else {
+                    deduped.set(order.price, deduped.get(order.price) + order.size);
+                }
+            });
+
+            const total = parsed.reduce((a, b) => ({
+                price: a.price + b.price,
+                size: a.size + b.size,
+            }));
+            return { parsed, total, deduped };
+        };
+
+        const orderBookSide = (orderBookForSide: LeafNode[], isBidSide?: boolean) => {
+            if (orderBookForSide) {
+                const _orderBookSide = getSide(orderBookForSide, isBidSide);
+                if (_orderBookSide) {
+                    return Array.from(_orderBookSide.deduped?.entries()).map((side) => [
+                        (side[0] / 10_000).toFixed(4),
+                        side[1],
+                    ]);
+                }
+            }
+            if (isBidSide) {
+                return [[0, 0]];
+            }
+            return [[69, 0]];
+        };
+
+        const getToB = (bids: LeafNode[], asks: LeafNode[]) => {
```

Comment:
An idea here would be to DRY some of this up and reference the subscription / needs of the orderbook component so we're not doing a ton of stuff we don't need or at least making the functions shared and available. I built this to get it working, not with the foresight of what I know now. Might be worth a poke to digest and review this such that these are common functions which will be abstracted out into a sdk regardless. Food for thought.